### PR TITLE
Refactor out queue from UI

### LINKED ISSUE #566: Refactor Qt widgets
Currently, many widgets don't follow the MVC pattern, nor use Qt's signals & slots. The biggest problem area from what I've seen is [`/ert_gui/simulation/detailed_progress.py`](https://github.com/equinor/ert/blob/f617112410d02cc8554d56c75e3cb896c82a8615/ert_gui/simulation/detailed_progress.py), where we should, among other things, use Qt's built-in widgets instead of calculating mouse-clicks and such ourselves.

This should be paired with #543 .

### LINKED ISSUE #1238: Remove any dependency on the queue system from the GUI/CLI
The GUI directly uses `JobQueueStatus` to perform various actions, e.g. like drawing colors. This is bad because the queue system _should_ be an implementation detail from the GUIs point of view. 

Secondly, the contract between the UIs and the experiment is messy. It currently consists of a `ForwardModelJobStatus` per forward model directly sourced from `status.json`, and a chunk of shared memory. The UIs control the experiment via `runSimulations` and `killAllJobs`. The latter causes problems like https://github.com/equinor/ert/issues/278 i.e. it is only possible to kill the queue, whereas the entire experiment should be killable.

So the contract should be
- some immutable data structure representing the status of the entire experiment (for now implemented in `SnapshotDict`)
- generic `play()` and `halt()` methods

### LINKED ISSUE #1288: Add constant strings for Waiting, Pending, Running, Failed, Finished, Unknown
And use the constants in place of the magic strings.

### PR DESCRIPTION
**Pull request by jondequinor**
**Issue**
Resolves #566 
Resolves #1238 
Resolves #1288

**Approach**

- Have snapshots drive UIs (GUI, CLI), forming a contract between UI code and the rest that does not rely on implementation details (queue statuses etc)
- Use proper `QAbstractItemModel` subclassing to drive the QT GUI
- Rewrite views so as to do less (i.e. never `update()`)

### CODE REVIEW

File: ert_shared/cli/monitor.py [hunk 1]
```diff
@@ -35,7 +37,8 @@ class Monitor(object):
 
     def __init__(self, out=sys.stdout, color_always=False):
         self._out = out
-
+        self._snapshot = None
+        self._start_time = None
```
Comment (jondequinor) [inline ert_shared/cli/monitor.py:41]: Yeah this can come from the event instead

File: ert_gui/simulation/view/realization.py [hunk 2]
```diff
@@ -38,26 +38,66 @@ def __init__(self, width, height, parent=None) -> None:
...
+            for _, step in stage.children.items():
+                # print("node2: ", step)
+                for _, job in step.children.items():
+                    # print("node3: ", job.data)
+                    status = job.data["status"]
+                    # print("job ", s)
+                    if status == "Success":
+                        status = "Finished"
+                    return QColor(*REAL_STATE_TO_COLOR[status])
+
```
Comment (jondequinor) [inline ert_gui/simulation/view/realization.py:54]: I'd really like to move this to the `data()` method of the snapshot model. The dumber these views are, the more are we conforming to the way of the QT.

Conversely, doing this in the model, allows you to later do

```python
painter.setBrush(self.model().data(index.row(), index, role=RealJobColorHint) 
```

Comment (frode-aarstad) [reply ert_gui/simulation/view/realization.py:54]: Agree, 100.001%

File: ert_gui/simulation/run_dialog.py [hunk 3]
```diff
@@ -197,21 +162,45 @@ def sizeHint(self) -> QSize:
...
+        self.setFixedHeight(200)
 
     def _setDetailedDialog(self) -> None:
         self._isDetailedDialog = True
         self._tab_widget.setVisible(True)
+        self._job_label.setVisible(True)
+        self._job_view.setVisible(True)
         self.show_details_button.setText("Hide Details")
         self.setFixedHeight(QWIDGETSIZE_MAX)
         self.setMinimumHeight(600)
```
Comment (jondequinor) [inline ert_gui/simulation/run_dialog.py:178]:
```suggestion
@Slot(QModelIndex, int, int)
```
It adds _some_ value both socially and technically, to indicate it's a slot.

File: ert_gui/model/progress_proxy.py [hunk 4]
```diff
@@ -1,9 +1,8 @@
 from qtpy.QtCore import QRect, QSize, QModelIndex, Qt, QIdentityProxyModel, QVariant
 
-#TODO: will these interfer with the ones in snapshot
-ProgressRole = Qt.UserRole + 1 
-SimpleProgressRole = Qt.UserRole + 2 
-
+# TODO: will these interfer with the ones in snapshot
```
Comment (jondequinor) [inline ert_gui/model/progress_proxy.py:3]: If you defined them in the snapshot module, then it might be less conflict-y. Then just import them here.

Comment (frode-aarstad) [reply ert_gui/model/progress_proxy.py:3]: Yes. It makes sense to collect the roles in the snapshot.

File: ert_gui/simulation/run_dialog.py [hunk 5]
```diff
@@ -200,31 +289,24 @@ def _on_ticker(self):
...
 
         if isinstance(event, EndEvent):
             self.simulation_done.emit(event.failed, event.failed_msg)
+            self._worker.stop()
+            self._ticker.stop()
+        elif isinstance(event, FullSnapshotEvent):
+            if event.snapshot is not None:
+                self._snapshot_model._add_snapshot(event.snapshot, event.iteration)
+            self._progress_view.setIndeterminate(event.indeterminate)
+            self._total_progress_bar.setValue(event.progress * 100)
```
Comment (sondreso) [inline ert_gui/simulation/run_dialog.py:301]: This triggers `DeprecationWarning: an integer is required (got type float).  Implicit conversion to integers using __int__ is deprecated, and may be removed in a future version of Python.` in Python 3.8. We should make the int conversion explicit.

File: ert_gui/simulation/run_dialog.py [hunk 6]
```diff
@@ -200,31 +289,24 @@ def _on_ticker(self):
...
+
+        elif isinstance(event, SnapshotUpdateEvent):
+            if event.partial_snapshot is not None:
+                self._snapshot_model._add_partial_snapshot(
+                    event.partial_snapshot, event.iteration
+                )
+            self._progress_view.setIndeterminate(event.indeterminate)
+            self._total_progress_bar.setValue(event.progress * 100)
```
Comment (sondreso) [inline ert_gui/simulation/run_dialog.py:309]: This triggers `DeprecationWarning: an integer is required (got type float).  Implicit conversion to integers using __int__ is deprecated, and may be removed in a future version of Python.` in Python 3.8. We should make the int conversion explicit.

File: ert_gui/simulation/view/progress.py [hunk 7]
```diff
@@ -0,0 +1,99 @@
...
+        self.background_color = QColor(200, 210, 210)
+
+    def paint(self, painter, option: QStyleOptionViewItem, index: QModelIndex) -> None:
+        data = index.data(ProgressRole)
+        if data is None:
+            return
+
+        nr_reals = data["nr_reals"]
+        status = data["status"]
+        d = option.rect.width() / nr_reals
```
Comment (sondreso) [inline ert_gui/simulation/view/progress.py:67]: This triggers `DeprecationWarning: an integer is required (got type float).  Implicit conversion to integers using __int__ is deprecated, and may be removed in a future version of Python.` in Python 3.8 (the trigger happens on line 89, when it is actually passed to the painter). We should make the int conversion explicit.

**Review by sondreso — COMMENTED** [review]
Some Python 3.8 warnings @frode-aarstad 😄

File: ert_shared/ensemble_evaluator/entity/ensemble.py [hunk 8]
```diff
@@ -642,7 +643,7 @@ def from_legacy(
                     .set_job_name(run_arg.job_name)
                     .set_run_arg(run_arg)
                 )
-        builder.set_metadata("iter", run_context.get_iter())
+        # builder.set_metadata("iter", run_context.get_iter())
```
Comment (sondreso) [inline ert_shared/ensemble_evaluator/entity/ensemble.py:646]: Remove?

File: ert_shared/ensemble_evaluator/evaluator.py [hunk 9]
```diff
@@ -1,4 +1,12 @@
 import asyncio
+from ert_shared.status.entity.state import (
+    ENSEMBLE_STATE_STARTED,
+    JOB_STATE_START,
+    REALIZATION_STATE_UNKNOWN,
```
Comment (sondreso) [inline ert_shared/ensemble_evaluator/evaluator.py:5]: Unused import?

File: ert_shared/status/tracker/evaluator.py [hunk 10]
```diff
@@ -0,0 +1,187 @@
...
+                self._work_queue.get_nowait()
+                self._work_queue.task_done()
+        except queue.Empty:
+            pass
+
+    def request_termination(self):
+        logger = logging.getLogger("ert_shared.ensemble_evaluator.tracker")
+
+        # There might be some situations where the
+        # evaulation is finished or the evaluation
```
Comment (sondreso) [inline ert_shared/status/tracker/evaluator.py:168]: Nitpick: evaluation

File: ert_shared/status/utils.py [hunk 11]
```diff
@@ -0,0 +1,103 @@
...
+    return "{byte_count:.2f} {suffix}".format(byte_count=byte_count, suffix=suffixes[i])
+
+
+def _calculate_progress(
+    finished: bool,
+    phase: int,
+    phase_count: int,
+    done_reals: int,
+    total_reals: int,
+    current_iter: int,
```
Comment (sondreso) [inline ert_shared/status/utils.py:24]: Parameter not needed?

File: ert_shared/status/utils.py [hunk 12]
```diff
@@ -0,0 +1,103 @@
...
+        tracker._model.currentPhase(),
+        tracker._model.phaseCount(),
+        done_reals,
+        total_reals,
+        current_iter,
+    )
+
+
+def format_running_time(runtime):
+    """ @rtype: str """
```
Comment (sondreso) [inline ert_shared/status/utils.py:53]: Type hinting instead

File: ert_shared/status/utils.py [hunk 13]
```diff
@@ -0,0 +1,103 @@
...
+        tracker.is_finished(),
+        tracker._model.currentPhase(),
+        tracker._model.phaseCount(),
+        done_reals,
+        total_reals,
+        current_iter,
+    )
+
+
+def format_running_time(runtime):
```
Comment (sondreso) [inline ert_shared/status/utils.py:52]: I know that this function probably isn't new in this PR, but isn't there better ways of doing this using datetime.timedelta?

Comment (sondreso) [reply ert_shared/status/utils.py:52]: To my suprise, that is actually not possible...

File: tests/cli/test_integration.py [hunk 14]
```diff
@@ -82,7 +82,9 @@ def test_runpath_file(tmpdir, source_root):
         assert os.path.isfile("RUNPATH_WORKFLOW_1.OK")
 
 
-def test_ensemble_evaluator(tmpdir, source_root):
+def test_ensemble_evaluator(tmpdir, source_root, caplog):
+    import logging
+    caplog.set_level(logging.DEBUG, logger="ert_shared.ensemble_evaluator")
```
Comment (sondreso) [inline tests/cli/test_integration.py:87]: I don't mind leaving the logging in, but if so we should move the import to the top of the file

File: tests/status/test_evaluator_tracker.py [hunk 15]
```diff
@@ -0,0 +1,100 @@
...
+
+def mock_wait_for_ws(*args, **kwargs):
+    raise ConnectionRefusedError("nah")
+
+
+def mock_ee_monitor(*args):
+    reals_ids = ["0", "1"]
+    snapshot = (
+        SnapshotBuilder()
+        .add_stage(stage_id="0", status="Running")
```
Comment (sondreso) [inline tests/status/test_evaluator_tracker.py:36]: This file has a lot of magic status strings

**Review by sondreso — COMMENTED** [review]
Think this PR looks really good, only small comments. There are some magic strings for statuses left in the evaluator, should we replace those also since this PR claims to resolve #1288

File: ert_shared/status/utils.py (hunk 13 above)
Comment (jondequinor) [reply ert_shared/status/utils.py:52]: Yes, no. I think I did some investigation the last time around. I will add a comment.

**Comment by sondreso** [comment]
There are also some magic status strings left in the evaluator: 
https://github.com/equinor/ert/blob/eb1920d8148969aba4b12bbc61dfaa88b4d1048b/ert_shared/ensemble_evaluator/evaluator.py#L105:L136

**Comment by jondequinor** [comment]
Yes, adressing them now. I do not address the magic strings in the config, nor the ones seemingly local to prefect (as I don't immediately know where to put them). Too far out of the scope of the bug in question.

**Review by sondreso — APPROVED** [review]

**Pull request merged**